From 79347866a5174c5941c4060c6fb97101efafd3a3 Mon Sep 17 00:00:00 2001
From: Jeremy Drake <sourceware-bugzilla@jdrake.com>
Date: Fri, 11 Sep 2020 17:51:16 +0100
Subject: [PATCH 5/8] Fix a segfault when creating an import library with 0
 exports.

	PR 26588
	* emultempl/pe.em (_finish): Only generate a import library if not
	exporting relocs.
	* emultempl/pep.em: Likewise.
---
 ld/ChangeLog        | 7 +++++++
 ld/emultempl/pe.em  | 3 ++-
 ld/emultempl/pep.em | 3 ++-
 3 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/ld/ChangeLog b/ld/ChangeLog
index f8ad16235f..06c06956a3 100644
--- a/ld/ChangeLog
+++ b/ld/ChangeLog
@@ -1,3 +1,10 @@
+2020-09-11  Jeremy Drake  <sourceware-bugzilla@jdrake.com>
+
+	PR 26588
+	* emultempl/pe.em (_finish): Only generate a import library if not
+	exporting relocs.
+	* emultempl/pep.em: Likewise.
+
 2020-08-28  Nick Clifton  <nickc@redhat.com>
 
 	PR 19011
diff --git a/ld/emultempl/pe.em b/ld/emultempl/pe.em
index f45518eafb..38b5536f72 100644
--- a/ld/emultempl/pe.em
+++ b/ld/emultempl/pe.em
@@ -2013,7 +2013,8 @@ gld_${EMULATION_NAME}_finish (void)
     )
     {
       pe_dll_fill_sections (link_info.output_bfd, &link_info);
-      if (command_line.out_implib_filename)
+      if (command_line.out_implib_filename
+          && pe_def_file->num_exports != 0)
 	pe_dll_generate_implib (pe_def_file, command_line.out_implib_filename,
 				&link_info);
     }
diff --git a/ld/emultempl/pep.em b/ld/emultempl/pep.em
index 7d9395168d..f161cce2b5 100644
--- a/ld/emultempl/pep.em
+++ b/ld/emultempl/pep.em
@@ -1823,7 +1823,8 @@ gld_${EMULATION_NAME}_finish (void)
 	  && pep_def_file->num_exports != 0))
     {
       pep_dll_fill_sections (link_info.output_bfd, &link_info);
-      if (command_line.out_implib_filename)
+      if (command_line.out_implib_filename
+          && pep_def_file->num_exports != 0)
 	pep_dll_generate_implib (pep_def_file,
 				 command_line.out_implib_filename, &link_info);
     }
-- 
2.28.0.windows.1

